swingset/src/main/java/com/nqadmin/swingset/SSBaseComboBox.java

```diff
@@ -332,6 +401,144 @@ public boolean hasSelection() {
+	        	getEditor().setItem(priorEditorText);   
+	        	//showPopup();// When this was working as intended, the ordering of showPopup() before updateUI() was relevant.
+	        	updateUI(); // NEEDED TO SHOW REVERTED ITEM (ORIGINALLY INTENDED TO SHOW REVERTED TEXT IN EDITOR)
+	        	
+	        	// WARN USER AND LOG
+	        	String editorText = getEditor().getItem().toString();
+	        	
+	        	JOptionPane.showMessageDialog(SSBaseComboBox.this,
+						"The text entered does not match any item in the list. Reverting to '" + editorText + "'.");
+	       	
+	        	logger.debug(() -> String.format("%s:   Editor string following revert action: '%s'.",
+	        			getColumnForLog(), editorText));
+
+	        } else {
+	        // SCENARIO #1-D ABOVE - SOMETHING ELSE NOT ANTICIPATED
+	        // SAME RESULT AS SCENARIO #3 BELOW
+	        	// TELL ACTION LISTENER NOT TO UPDATE ITEM OR ROWSET
+	        	actionListenerNoUpdate = true; // TELL ACTION LISTENER NOT TO UPDATE ITEM OR ROWSET
+	        	
+	        	// WARN USER AND LOG
+	        	JOptionPane.showMessageDialog(SSBaseComboBox.this, String.format(
+						"Unexpected call to setSelectedItem() for column %s with value '%s'.", getColumnForLog(), _value));
+	        	logger.warn(() -> String.format("%s: Unexpected call to setSelectedItem() with '%s'.)",
+	        			getColumnForLog(), _value));
+	        }
+		} else if (_value instanceof SSListItem) {
+		// #2 - WE HAVE A SSLISTITEM (including nullItem if getAllowNull()==true)
+			super.setSelectedItem(_value);
+			priorEditorText = currentEditorText;
+			// 2020-12-29_BP: CONFIRMED selectAll() IS NEEDED FOLLOWING A NORMAL ITEM SELECTION
+			//   OTHERWISE USER IS APPENDING EXISTING ITEM STRING IF THEY START TO TYPE
+			if (getEditor() != null && getEditor().getEditorComponent().hasFocus()) {
+				// after we find a match, do a select all on the editor so
+				// if the user starts typing again it won't be appended
+				// 2020-12-21_BP: if we don't limited to field with focus, the comboboxes blink on navigation
+				// this also causes the focus to jump out of a navigation combo
+				getEditor().selectAll();
+			}
+			logger.debug("{}: Valid match. Selected item after super.setSelectedItem()={}", () -> getColumnForLog(), () -> getSelectedItem());
+		} else {
+	    // SCENARIO #3 - WE HAVE A STRING OR SOME OBJECT OTHER THAN NULL OR SSLISTITEM
+		// SAME RESULT AS SCENARIO #1-D ABOVE
+        	// TELL ACTION LISTENER NOT TO UPDATE ITEM OR ROWSET
+        	actionListenerNoUpdate = true; // TELL ACTION LISTENER NOT TO UPDATE ITEM OR ROWSET
+        	
+        	// WARN USER AND LOG
+        	JOptionPane.showMessageDialog(SSBaseComboBox.this, String.format(
+					"Unexpected call to setSelectedItem() for column %s with value '%s'.", getColumnForLog(), _value));
+        	logger.warn(() -> String.format("%s: Unexpected call to setSelectedItem() with '%s'.)",
+        			getColumnForLog(), _value));
+			
+		}
+	}
```

Comment:
I agree. I don't see how it should land in this else. Hopefully it doesn't, but based on other strangeness, I thought it was good to have a catch all and bring it to the user's attention (and hopefully to our attention).

Reply:
There are 2 of roughly 6 cases where we want this set to true. Seems cleaner to me to initialize to false and then set to true in 2 cases where it's needed rather than introducing break/return in 4 places to prevent it from being set to false. Let's discuss further by email if needed. Maybe there is a best practice I'm missing/convention I'm missing/violating.

Reply:
I was rolling this around in my head some more. You can ignore the above comment about setting to true for 2 of the 6 cases. I didn't get that you were just talking about clean up on exiting the method. It probably could be set to false upon leaving setSelectedItem(), but I don't think it has to be. Will get reset to false upon the next call to setSelectedItem() and the Action Listener shouldn't do anything in the interim.